src/AutomaticScheduling.cpp

```diff
@@ -0,0 +1,268 @@
+        return UNDEFINED_FOOTPRINT_SIZE;
+    }
+    Expr footprint = 1;
+    for (vector<Interval>::iterator I = bounds.begin(), E = bounds.end(); I != E; ++I) {
+        const Interval &i = *I;
+        internal_assert(i.min.defined() && i.max.defined());
+        Expr diff = i.max - i.min + 1;
+        footprint *= diff;
+    }
+    const int *result = as_const_int(simplify(footprint));
+    if (result) {
+        return (unsigned)*result;
+    } else {
+        return UNDEFINED_FOOTPRINT_SIZE;
+    }
+}
+
+} // end anonymous namespace
+
+void ComputeRootAllStencils::apply(Func root) {
+    // Construct a callgraph for the pipeline.
+    CallGraph cg(root.function());
+    vector<Function> all_functions = cg.transitive_calls(root.function());
+    for (vector<Function>::iterator I = all_functions.begin(), E = all_functions.end(); I != E; ++I) {
+        Function f = *I;
+        unsigned footprint = calculate_footprint_size(f, cg);
+        if (footprint > 1) {
+            Func wrapper(f);
+            wrapper.store_root().compute_root();
+        }
+    }
+}
+
+void ParallelizeOuter::apply(Func root) {
+    CallGraph cg(root.function());
+    vector<Function> all_functions = cg.transitive_calls(root.function());
+    all_functions.push_back(root.function());
+    for (vector<Function>::iterator I = all_functions.begin(), E = all_functions.end(); I != E; ++I) {
+        Function f = *I;
+        if (!f.schedule().compute_level().is_inline()) {
+            Func wrapper(f);
+            Dim outer = f.schedule().dims()[f.schedule().dims().size() - 1];
+            Var v(outer.var);
+            wrapper.parallel(v);
+        }
+    }
+}
+
+void VectorizeInner::apply(Func root) {
+    CallGraph cg(root.function());
+    vector<Function> all_functions = cg.transitive_calls(root.function());
+    all_functions.push_back(root.function());
+    for (vector<Function>::iterator I = all_functions.begin(), E = all_functions.end(); I != E; ++I) {
+        Function f = *I;
+        if (!f.schedule().compute_level().is_inline()) {
+            Func wrapper(f);
+            Dim inner = f.schedule().dims()[0];
+            Var v(inner.var);
+            unsigned factor = 128 / f.output_types()[0].bits;
+            wrapper.vectorize(v, factor);
+        }
+    }
+}
+
+void apply_automatic_schedule(Func root, AutoScheduleStrategy strategy, bool reset_schedules) {
+    AutoScheduleStrategyImpl *impl = NULL;
+    switch (strategy) {
+    case AutoScheduleStrategy::ComputeRootAllStencils:
+        impl = new ComputeRootAllStencils();
+        break;
+    case AutoScheduleStrategy::ParallelizeOuter:
+        impl = new ParallelizeOuter();
+        break;
+    case AutoScheduleStrategy::VectorizeInner:
+        impl = new VectorizeInner();
+        break;
+    }
+    internal_assert(impl != NULL);
+    // Reset all user-specified schedules.
+    if (reset_schedules) ResetSchedules reset(root.function());
+    impl->apply(root);
```

Comment:
Please let's officially flip that bit; unique_ptr alone makes it all worthwhile :-)